### ISSUE TITLE
Issues upgrading to Next 9

### ISSUE THREAD

**Issue by hstove** [zeit/next.js#8775]
# Bug report

First off, I want to thank the Zeit team and contributors for next.js. I work at @blockstack, and we use Next.js for all of our web apps. We're big fans.

I just upgraded one of our next.js apps from v8.0.x to v9.0.5. It was more of a hassle than previous upgrades, and I wanted to outline some of the issues that I ran into. Maybe some of these can be fixed, and I am hoping that this can at least demonstrate some workarounds for anyone Googling some of these errors.

The app is https://app.co, which is [open source on Github](https://github.com/blockstack/app.co). It's 100% javascript (no TypeScript), and uses a custom server.

I have successfully upgraded the app to v9 after fixing these issues, and you can view the [Pull Request](https://github.com/blockstack/app.co/pull/628) containing that work to follow along.

### 1. `TypeError: Property name expected type of string but got null`

After updating my `package.json` to next 9 and running `yarn`, this was the first issue I ran into when trying to run the app. The actual error was `Cannot find module "pages/_document"`, but that was clearly not the root of the issue. We already had next.config.js and webpack configured to fix some `resolve.alias` issues that others run into. Ultimately, it seemed to come down to the `TypeError` pasted above. There was a 1-line backtrace, but only indicating the file where the error was found, without any specific lines.

This really stumped me, especially because the app had no TypeScript. I ended up having to comment out different parts of the file until I was able to nail down the error. **The error was due to object destructuring**. It doesn't seem to throw for all destructure calls, but it does consistently for some, like `const { apps } = await response.json()` when using `fetch`. Additionally, I had to fix all destructure calls in `getInitialProps`.

I ended up working around this by re-writing those lines to not destructure. It's uglier, but it compiles. [Here is a commit that demonstrates some of the changes I had to make](https://github.com/blockstack/app.co/pull/628/commits/ec7791fb1783833a070b7304e9fed8a51f34b6d2).

After fixing all of the compile errors related to destructuring, the app would compile and run successfully.

I am interested in hearing why this error came up in the first place. I would assume that if I re-wrote those files in TS, I would get destructuring working, assuming it was type safe. It would also be nice to have a better backtrace for type errors.

### 2. I still had to use `@zeit/with-typescript` to successfully get a TypeScript component to resolve and compile.

We want to start migrating this project to TS, so I wanted to see if it was as easy as just changing a file to `.tsx`. There were a few type errors, which were valid. I fixed those, but then when I would visit that page, it would 404 with no error message that I could see. (I was changing a component in the `pages` directory).

Ultimately, I added `@zeit/next-typescript` to `next.config.js`, and page finally rendered as expected. I get a warning when I run the app that I don't need it, but at least it works.

You can see the work that got this to run [in this commit](https://github.com/blockstack/app.co/pull/628/commits/9f6d7128f118bd5964286a75790604feb3160d57).

This was also after running `next dev`, which properly detected TS and setup our `tsconfig.json` for me. By the way, it's not clear that you have to change a component in `/pages` for this to happen, because at first I tried to change a component in a different folder. I found out that next only detects `/pages` components by reading the changelog and finding #8479. I'd suggest adding that to the v9 blog post.

## To Reproduce

You can reproduce this by running [app.co](https://github.com/blockstack/app.co). The current master branch is still on v8. Alternatively, you can checkout `539b14b`, which is the commit that only bumps `next.js` to v9, but doesn't have any fixes, so you can repro the bug on that commit. Note - you'll need redis running on your machine for the app to run.

You can run the app with `node server.js`, and you should see the first issue listed above.

## Expected behavior

Since the app is 100% javascript, I was surprised to see type errors. Additionally, I am unsure why I still need to use `@zeit/next-typescript`.

## System information

- OS: macOS
- Version of Next.js: 9.0.5
- node.js 10.15.3

**Comment by hstove** [comment]
Follow up - the reason I had the second issue is because I was overriding the `pageExtensions` option in `next.config.js`. I ended up removing that, because it was taken care of by `@next/mdx` anyways.

**Comment by Timer** [comment]
To clarify, a [`TypeError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError) is not from TypeScript -- this is a standard JavaScript error.

The two (TypeError and TypeScript) seem to have been conflated in your report. Also, the [link you provided](https://github.com/blockstack/app.co/pull/628/commits/11b7d3e4227351a4214dc378a1ab7e11708ba3660) is broken.

Can you please show us code that reproduces this? Object destructuring is expected to work.

**Comment by hstove** [comment]
Apologies for conflating that. Here is one of the commits that fixed some of the compilation errors I ran into related to destructuring: https://github.com/blockstack/app.co/pull/628/commits/ec7791fb1783833a070b7304e9fed8a51f34b6d2

**Comment by hstove** [comment]
I figured out the root cause of the issue. I noticed that the destructuring errors only happened in `async` functions. I figured this had to do with something Babel-related. We had `@babel/plugin-transform-regenerator` and `@babel/plugin-transform-runtime` in our `.babelrc`, which we added a while ago, as this is a somewhat older Next.js project. I removed them from `.babelrc`, assuming that's all in `next/babel` now. The project is running smoothly and that `TypeError` is gone.
